okhttp/src/main/java/com/backblaze/b2/client/okHttpClient/B2OkHttpClientImpl.java

```diff
@@ -0,0 +1,327 @@
+            } else {
+                if( response.code() == 200){
+                    Headers responseHeaders = response.headers();
+                    return makeB2Headers(responseHeaders);
+                } else {
+                    throw B2Exception.create(null, response.code(), null, "");
+                }
+            }
+        } catch (IOException e) {
+            throw translateToB2Exception(e, url);
+        }
+    }
+
+    @Override
+    public void close() {
+        // Closing response body streams is handled elsewhere by OkHttp.
+    }
+
+    private String postJsonAndReturnString(String url,
+                                           B2Headers headersOrNull,
+                                           Object request) throws B2Exception {
+        try {
+            B2Json bzJson = B2Json.get();
+            String requestJson = bzJson.toJson(request);
+            byte[] requestBytes = getUtf8Bytes(requestJson);
+            return postAndReturnString(url, headersOrNull,new ByteArrayInputStream(requestBytes), requestJson.length());
+        } catch (B2JsonException e) {
+            //log.warn("Unable to serialize " + request.getClass() + " using B2Json, was passed in request for " + url, ex);
+            throw new B2LocalException("parsing_failed", "B2Json.toJson(" + request.getClass() + ") failed: " + e.getMessage(), e);
+        }
+
+    }
+
+    private byte[] readFully(InputStream inputStream, int contentLength) throws IOException{
+        int numRead = 0;
+        int offset = 0;
+        int length = contentLength;
+        byte[] bytes = new byte[contentLength];
+        while( numRead < contentLength ){
+            int n = inputStream.read(bytes, offset, length);
+            if( n > -1 ){
+                numRead += n;
+                offset += n;
+                length = contentLength - numRead;
+            }
+        }
+        return bytes;
+    }
+
+    /**
+     * POSTs to a web service that returns content, and returns the content
+     * as a single string.
+     *
+     * @param url the url to post to
+     * @param headersOrNull the headers, if any.
+     *
+     * @return the body of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    private  String postAndReturnString( String url,  B2Headers headersOrNull,  InputStream inputStream, long contentLength)
+            throws B2Exception {
+        RequestBody body;
+        try {
+            if( contentLength <= (Integer.MAX_VALUE-HOTSPOT_FUDGE) ){
+                byte[] bytes = readFully(inputStream, (int) contentLength);
+                body = RequestBody.create(MediaType.get(APPLICATION_JSON), bytes);
+            } else {
+                String tempFileName = UUID.randomUUID().toString();
+                File tempFile = File.createTempFile(tempFileName, ".tmp");
+                FileOutputStream outputStream = new FileOutputStream(tempFile);
+                copy( inputStream, outputStream);
```

Comment:
copy() doesn't close the destination stream.  i think you need to do that to make sure everything has made it to disk.

Reply:
Is there a way to open the stream using a try with resources that way we don't need to remember to close it?
Just a thought, might not be a solution.

Reply:
@OldSkoolMark -- do you need to close the stream here?

Reply:
now closing immediately after the copy